File: tests/cpp/test_triangle7.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//

/**
 * @file test_triangle7.cpp
 * @brief `triangle7` (v16.0.0): a `triangle6` with a centre node, as Code_Aster
 *        (`TRIA7`), MED (`TR7`) and VTK (34) define it. The cell-type tables,
 *        the edge table, the corner count and the linear base.
 */

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/cell_type.hpp"
#include "meshioplusplus/detail/cell_edges.hpp"
#include "meshioplusplus/detail/geometry.hpp"
#include "meshioplusplus/operations/convert_cells.hpp"
#include "meshioplusplus/operations/surface.hpp"
#include "meshioplusplus/types.hpp"

using namespace meshioplusplus;

namespace {

Mesh triangle7_mesh() {
    return mt::make_mesh({{0, 0, 0},
                          {1, 0, 0},
                          {0, 1, 0},
                          {0.5, 0, 0},
                          {0.5, 0.5, 0},
                          {0, 0.5, 0},
                          {1.0 / 3.0, 1.0 / 3.0, 0}},
                         "triangle7", {{0, 1, 2, 3, 4, 5, 6}});
}

}  // namespace

TEST(Triangle7, IsAKnownSevenNodeSurfaceCell) {
    const CellType t = cell_type_from_name("triangle7");
    ASSERT_EQ(t, CellType::Triangle7);
    EXPECT_EQ(cell_type_name(t), "triangle7");
    EXPECT_EQ(cell_type_num_nodes(t), 7);
    EXPECT_EQ(cell_type_dimension(t), 2);
    EXPECT_EQ(detail::cell_corner_count(t), 3);
}

TEST(Triangle7, HasTheEdgesOfATriangle6) {
    const auto& t7 = detail::cell_edges(CellType::Triangle7);
    const auto& t6 = detail::cell_edges(CellType::Triangle6);
    ASSERT_EQ(t7.size(), 3u);
    for (std::size_t i = 0; i < t7.size(); ++i) {
        EXPECT_EQ(t7[i].mEdgeType, t6[i].mEdgeType);
        EXPECT_EQ(t7[i].mNodes, t6[i].mNodes);
    }
}

TEST(Triangle7, LinearizesToATriangle) {
    const ConvertCellsResult r = convert_cells(triangle7_mesh());
    ASSERT_EQ(r.mMesh.NumCellBlocks(), 1u);
    EXPECT_EQ(r.mMesh.Cells(0).Type(), "triangle");
    EXPECT_EQ(r.mMesh.NumPoints(), 3u);
}

TEST(Triangle7, SurfaceIsItsThreeQuadraticEdges) {
    const Mesh s = extract_surface(triangle7_mesh());
    ASSERT_EQ(s.NumCellBlocks(), 1u);
    EXPECT_EQ(s.Cells(0).Type(), "line3");
    EXPECT_EQ(s.Cells(0).NumCells(), 3u);
}

File: tests/python/test_triangle7.py
"""``triangle7`` (v16.0.0): a ``triangle6`` plus a centre node, as Code_Aster
(``TRIA7``), MED (``TR7``) and VTK (34) define it. Before it was a cell type the
VTK and MED readers named blocks ``triangle7`` that nothing else could size."""

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus._common import num_nodes_per_cell
from meshioplusplus.med import _med as py_med
from meshioplusplus.vtu import _vtu as py_vtu

POINTS = np.array(
    [
        [0.0, 0.0, 0.0],
        [1.0, 0.0, 0.0],
        [0.0, 1.0, 0.0],
        [0.5, 0.0, 0.0],
        [0.5, 0.5, 0.0],
        [0.0, 0.5, 0.0],
        [1 / 3, 1 / 3, 0.0],
    ]
)


def _mesh():
    return meshioplusplus.Mesh(POINTS, [("triangle7", [list(range(7))])])


def test_is_a_seven_node_cell():
    assert num_nodes_per_cell["triangle7"] == 7
    assert _mesh().cells[0].dim == 2


@pytest.mark.parametrize(
    "fmt, core, python",
    [
        ("vtu", meshioplusplus.vtu, py_vtu),
        ("med", meshioplusplus.med, py_med),
    ],
)
@pytest.mark.parametrize("engine", ["core", "python"])
def test_round_trips(tmp_path, fmt, core, python, engine):
    if fmt == "med":
        pytest.importorskip("h5py")
    module = core if engine == "core" else python
    path = tmp_path / f"t7.{fmt}"
    module.write(path, _mesh())
    back = meshioplusplus.read(path)
    assert [b.type for b in back.cells] == ["triangle7"]
    np.testing.assert_array_equal(back.cells[0].data, [list(range(7))])
    np.testing.assert_allclose(back.points[:, :2], POINTS[:, :2])


def test_linearizes_to_a_triangle():
    out = meshioplusplus.convert_cells(_mesh(), mode="linearize")
    assert [b.type for b in out.cells] == ["triangle"]
    assert len(out.points) == 3


def test_surface_is_three_quadratic_edges():
    surface = meshioplusplus.extract_surface(_mesh())
    assert [(b.type, len(b.data)) for b in surface.cells] == [("line3", 3)]
